### DISCUSSION TITLE
[CHORE] 如何配置多个webhooks.WebhookOptions选项？

### DISCUSSION THREAD

**Discussion by zqhgit001** [elsa-workflows/elsa-core#5479] [Q&A]
您好：如何配置多个webhooks.WebhookOptions？

**Comment by glime-ai** [comment]
您好，要配置多个 `webhooks.WebhookOptions`，您需要在 `appsettings.json` 文件中定义多个 Webhook 端点配置，并在启动时将这些配置绑定到 Elsa 的 Webhook 选项中。以下是一个示例，展示了如何在 `appsettings.json` 中配置多个 Webhook 端点，并在启动时绑定这些配置：

1. 在 `appsettings.json` 中添加多个 Webhook 端点配置：
```json
{
  "Webhooks": {
    "Endpoints": [
      {
        "EventTypes": ["EventType1"],
        "Url": "https://your-webhook-url-1"
      },
      {
        "EventTypes": ["EventType2"],
        "Url": "https://your-webhook-url-2"
      }
      // 可以继续添加更多的 Webhook 端点配置
    ]
  }
}
```

2. 在启动时绑定 Webhook 配置：
```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddElsa(elsa =>
{
    // 其他 Elsa 配置...

    // 绑定 Webhook 配置
    elsa.UseWebhooks(webhooks => webhooks.WebhookOptions = options => builder.Configuration.GetSection("Webhooks").Bind(options));
});
```

通过这种方式，您可以在 `appsettings.json` 中定义多个 Webhook 端点，并在应用程序启动时将它们绑定到 Elsa 的 Webhook 选项中。这样，Elsa 就能够根据配置处理多个 Webhook。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by glime-ai** [comment]
您好，要配置多个 `webhooks.WebhookOptions`，您需要在 `appsettings.json` 文件中定义多个 Webhook 端点，并在启动配置中绑定这些设置。以下是一个示例，展示了如何在 `appsettings.json` 中配置多个 Webhook 端点：

```json
{
  "Webhooks": {
    "Endpoints": [
      {
        "EventTypes": ["EventType1"],
        "Url": "https://example.com/webhooks/endpoint1"
      },
      {
        "EventTypes": ["EventType2"],
        "Url": "https://example.com/webhooks/endpoint2"
      }
      // 可以继续添加更多的 Webhook 端点配置
    ]
  }
}
```

然后，在您的启动配置（例如 `Program.cs` 或 `Startup.cs`）中，您需要使用 `UseWebhooks` 扩展方法来绑定这些设置：

```csharp
builder.Services.AddElsa(elsa =>
{
    // 其他配置...

    // 注册自定义 Webhook 定义
    elsa.UseWebhooks(webhooks => webhooks.WebhookOptions = options =>
        builder.Configuration.GetSection("Webhooks").Bind(options));
});
```

这样，Elsa 就会根据您在 `appsettings.json` 中定义的配置来注册多个 Webhook 端点。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by YangYanchengCN** [comment]
最新版本的代码，webhooks.WebhookOptions 应为 webhooks.ConfigureSinks：
//webhooks.WebhookOptions = options =>
//{
//    builder.Configuration.GetSection("Webhooks").Bind(options);
//};
 webhooks.ConfigureSinks = options =>
 {
     builder.Configuration.GetSection("Webhooks").Bind(options);
 };

**Comment by sfmskywalker** [comment]
The current API has changed from the older examples.

In the current codebase, `UseWebhooks(...)` exposes `ConfigureSinks`, not `WebhookOptions`:

- feature surface: `/Users/sipke/Projects/Elsa/elsa-extensions/src/modules/http/Elsa.Http.Webhooks/Features/WebhooksFeature.cs`
- current sample usage: `/Users/sipke/Projects/Elsa/elsa-samples/src/aspnet/Elsa.Samples.AspNet.Webhooks.WorkflowServer/Program.cs`
- another current sample with `Webhooks:Sinks`: `/Users/sipke/Projects/Elsa/elsa-samples/src/aspnet/Elsa.Samples.AspNet.Onboarding.WorkflowServer/Program.cs`

So for current versions, the configuration pattern is:

```csharp
elsa.UseWebhooks(webhooks =>
    webhooks.ConfigureSinks = options =>
        builder.Configuration.GetSection("Webhooks").Bind(options));
```

And the config shape contains a `Sinks` collection, e.g.:

```json
"Webhooks": {
  "Sinks": [
    {
      "Id": "ExternalApplication.RunTask",
      "Name": "Run Task",
      "EventTypes": ["RunTask"],
      "Url": "https://localhost:6001/api/webhooks/run-task"
    }
  ]
}
```

So if your question is "how do I configure multiple webhook destinations?", the answer is: add multiple entries to `Webhooks.Sinks` and bind them through `ConfigureSinks`.

One caveat: there is still an older guide in `elsa-guides/src/guides/external-app-interaction/elsa-server/ElsaServer/Program.cs` that uses `webhooks.WebhookOptions`. That guide is older than the current feature surface, while `elsa-guides/src/installation/elsa-server/ElsaServer/Program.cs` already uses `ConfigureSinks`.

So the latest community comment pointing to `ConfigureSinks` is the correct direction for current code.
